File: [6]Zigzag_Conversion.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1)
            return s;

        string result[numRows];
        string final_str = "";
        int inc = 0;
        int inc_amount = 1;

        for (int i = 0; i < s.length(); ++i){
            result[inc] += s[i];
            if ((inc == 0) and (inc_amount == -1)){
                inc_amount *= -1;
            } else if ((inc == numRows-1) && (inc_amount == 1)){
                inc_amount *= -1;
            }
            inc += inc_amount;
        }

        for (int i=0; i<numRows; i++){
            for (int j=0; j< result[i].length() ; j++)
                final_str += result[i][j];
        }

        return final_str;
    }
};

int main(void){
    int rows = 3;
    string inputs[rows][3] = {
        {"ABCD", "2", "ACBD"},
        {"PAYPALISHIRING", "3","PAHNAPLSIIGYIR"},
        {"ABCDE", "3" ,"AEBDC"}
    };

    for (int i=0; i<rows; i++){
        if ( inputs[i][2] == Solution().convert(inputs[i][0], stoi(inputs[i][1])) ){
            cout << "Correct" << endl;
        } else
            cout << "Incorrect Answer" << endl;
        cout << endl;
    }
    return 0;
}

File: [6]Zigzag_Conversion.py
class Solution:
    def convert(self, s: str, numRows: int):
        if numRows == 1:
            return s
        result = [[] for i in range(numRows)]
        isIncrement = True
        inc = 0
        for ch in s:
            if (inc == -1) or (inc == numRows):
                isIncrement = False if isIncrement else True
                inc = inc+2 if isIncrement else inc-2
            result[inc].append(ch)
            inc = inc+1 if isIncrement else inc-1

        ans = ""
        for res in result:
            for r in res:
                ans += r
        return ans

# Do the fastest solution from leetcode here.
# class Solution:
#     def convert(self, s: str, numRows: int):
#         str = [""] *numRows
#         isInc = True
#         inc = 1
#         i = 0
#         for ch in s:
#             str[i] += ch
#             pass

inputs = [
    ["PAYPALISHIRING", 3,"PAHNAPLSIIGYIR"],
    ["ABCDE", 4 ,"ABCED"]
]

for original_string, numRows, solution_String in inputs:
    if ( (solution_String) != (result := Solution().convert(original_string, numRows)) ):
        print("Wrong Answer  => ", result)